### DISCUSSION TITLE
Repository working flow

### DISCUSSION THREAD

**Discussion by tai271828** [pycontw/pycon-etl#15] [General]
When working on https://github.com/pycontw/PyCon-ETL/issues/5 , we started the discussion thread regarding which working flow we may want to use in [this pull request](https://github.com/pycontw/PyCon-ETL/pull/6). For example, "git flow", "github flow", or "gitlab flow". The most impact is that we are running docker-registry-based CI tasks: what if we use some workflow, and we may want to apply the docker registry credential? Some thinking points of view pop up in my mind:

1. Re-visit the usage of docker tests. Are they necessary to run in every scenarios?
2. Figure out the current deployment policy
3. How to integrate pull requests from community?

...and more.

**Comment by tai271828** [comment]
By the way I was aware of [the squashed commit](https://github.com/pycontw/PyCon-ETL/commit/981ca82a116b8dd1ad7b7f07dbd3c75ce95cd561) in the master branch recently, and I don't think this kind of squashing is suggested in the forementioned "git flow", "github flow", or "gitlab flow", is it?

**Reply by Lee-W** [reply]
I suggest against using squashing as well. It would make tracing harder in the future as "squashing everything" might not be a readable enough comment.

**Reply by david30907d** [reply]
agree with squashing in the master branch.

I did this is because @tai271828 mentioned about `questionnaire` data was uploaded to Github. Although there's no confidential data in it, still don't feel comfortable to have it in our git log 😅 

sorry for the inconvenience

**Reply by tai271828** [reply]
> agree with squashing in the master branch.
> 
> I did this is because @tai271828 mentioned about `questionnaire` data was uploaded to Github. Although there's no confidential data in it, still don't feel comfortable to have it in our git log 
> 
> sorry for the inconvenience

I see. Thanks for clarification.

**Comment by Lee-W** [comment]
If I'm not misunderstood, what we're using now is more like "GitHub flow."

1. If the cost is not that high, I think it's still better to run on every pull request.
2. no idea on this one, haha. maybe @david30907d would know
3. I'm not sure I understand this one? Could you please provide more detail?

**Reply by david30907d** [reply]
0. well, i'm using `gitlab flow` actually ([ref](https://medium.com/@lf2lf2111/%E4%B8%89%E7%A8%AE%E7%89%88%E6%8E%A7%E6%B5%81%E7%A8%8B-29c82f5d4469)
1. agree with @Lee-W , go for running on every PR
2. Just to clarify, `deployment policy` means when and how do i deploy this repo? Manually lol
3. Prefer gitlab flow so that i can apply the docker registry credential. Okay with `github flow` if we can figure out how to setup the credential in this kind of scenario~

**Reply by tai271828** [reply]
> 1. well, i'm using `gitlab flow` actually ([ref](https://medium.com/@lf2lf2111/%E4%B8%89%E7%A8%AE%E7%89%88%E6%8E%A7%E6%B5%81%E7%A8%8B-29c82f5d4469)
> 2. agree with @Lee-W , go for running on every PR
> 3. Just to clarify, `deployment policy` means when and how do i deploy this repo? Manually lol
> 4. Prefer gitlab flow so that i can apply the docker registry credential. Okay with `github flow` if we can figure out how to setup the credential in this kind of scenario~

Regarding item0 and item2:
1) If we are using `gitlab flow`, should I expect there is `staging` branch and `production` branch? <--- this question is for @david30907d  : )
2) What I meant by `deployment policy` is that I expect we have branches like `staging` and `production` and deploy based on the corresponding branches.

Regarding item3:
@david30907d is right. If we use `github flow`, and we want to run CI for every pull requests to master, we need to think about the "docker image" test for the lack of credential. @Lee-W  please refer to this failure for more details.

Currently we (me, david, wei) have permission to push branches to THIS repository, and this will ensure our CI triggered by our pull requests could run normally. However for people without permission to push branches to this repository, the corresponding is expected to fail certainly for the lack of docker image credential. This may block the pull requests from people without permissions.
